File: typescript/F6_objects.ts
// ============================================
// F6 - OBJECTS / DICTIONARIES
// Project: VoiceDoc AI - Doctor Voice Agent
// Run: bun typescript/F6_objects.ts
// ============================================


// ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
// PART 1 - Type Define Karo (TS Only)
// ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
 
 
 // Doctor ka type — har field ka type batao


 type Doctor ={
    name : string , 
    id : string , 
    specialization : string , 
    fee : number ,
    rating : number , 
    available : boolean, 
    address :{
        citys : string, 
        room : string, 
        hospital : string 

    }
 };


 // Session ka type
type Session = {
    doctorNam: string;
    patientNam: string;
    feed: number;
    duration: number;
    transcript: string[];
};  


// ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
// PART 2 - Object Banana
// ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━


const DoctorProfile : Doctor =   {
    id: "DOC001",
    name: "Dr. Priya Sharma",
    specialization: "Cardiology",
    fee: 800,
    rating: 4.8,
    available: true,
    address: {
        hospital: "Apollo Hospital",
        citys: "Delhi",
        room: "101"
    }
};

console.log("\n--- Doctor Profile ---");

console.log(DoctorProfile);

DoctorProfile.fee =1000
console.log(DoctorProfile)

// ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
// PART 3 - Access Karna
// 

// accesing values 

console.log(DoctorProfile.address.room); //// Nested access
console.log("Fee:", DoctorProfile.fee); // dot notation 


// ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
// PART 4 - Destructuring
// ━━ 

// ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
// PART 4 - Destructuring
// ━━
// ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
// PART 4 - Destructuring
// ━━
// ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
console.log("\n--- Destructuring ---");

const sesion : Session ={

     doctorNam: "Dr. Priya Sharma",
    patientNam: "Ramesh Kumar",
    feed: 800,
    duration: 30,
    transcript: []
}

// baisc desruction

const {doctorNam, patientNam , feed} =sesion ;

console.log("doctorname " , doctorNam);
console.log("patient name " , patientNam);
console.log("fees" , feed)

// Nested destructuring 

const {address :{citys , hospital}} =DoctorProfile;

console.log("City:", citys);
console.log("Hospital:", hospital);

// ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
// PART 5 - Spread Operator
// ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

console.log("\n--- Spread Operator ---");

// Session shuru — doctor busy ho gaya
const activeSession = {
    ...DoctorProfile,       // Sab copy karo
    available: false,       // Override karo
    currentPatient: "Ramesh Kumar",
    sessionStart: "10:30 AM"
};

console.log("Available:", activeSession.available);   // false
console.log("Patient:", activeSession.currentPatient);
console.log("Original Available:", DoctorProfile.available); // true — safe!

// ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
// PART 6 - Object Methods + Loop
// ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
console.log("\n--- Object Loop ---"); 

// Object.entries() se loop

for (const [key , value ] of Object.entries(DoctorProfile)){
    if (typeof value !== "object"){
        console.log(`${key}: ${value}`);
    }
}

// ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
// PART 7 - Real Project Use
// ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

const { name: docName, specialization, fee: docFee } = DoctorProfile;

console.log(`
╔══════════════════════════════╗
║      VOICEDOC AI SESSION     ║
╠══════════════════════════════╣
║ Doctor : ${docName}
║ Dept   : ${specialization}
║ Fee    : ₹${docFee}
║ Patient: ${activeSession.currentPatient}
║ Time   : ${activeSession.sessionStart}
║ Status : ${activeSession.available ? "Free" : "In Session"}
╚══════════════════════════════╝
`);

File: python/F6_objects.py
# ============================================
# F6 - OBJECTS / DICTIONARIES
# Project: VoiceDoc AI - Doctor Voice Agent
# Run: python python/F6_objects.py
# ============================================


# ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
# PART 1 - Dictionary Banana
# ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
doctor_profile = {
    "id": "DOC001",
    "name": "Dr. Priya Sharma",
    "specialization": "Cardiology",
    "fee": 800,
    "rating": 4.8,
    "available": True,
    "address": {
        "hospital": "Apollo Hospital",
        "city": "Delhi",
        "room": 101
    }
}

print("\n--- Doctor Profile ---")
print(doctor_profile)

# ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
# PART 2 - Access Karna
# ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
print("\n--- Accessing Values ---")
# Bracket notation — Python ka tarika
print("Name:", doctor_profile["name"])
print("Fee:", doctor_profile["fee"])


# .get() — safe tarika
print("Phone:", doctor_profile.get("phone", "N/A"))  # N/A — error nahi!


# Nested access
print("Hospital:", doctor_profile["address"]["hospital"])
print("City:", doctor_profile["address"]["city"])


# ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
# PART 3 - Modify Karna
# ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

print("\n--- Modifying ---")

# Update
doctor_profile["fee"] = 1000
print("Updated Fee:", doctor_profile["fee"])

# Add
doctor_profile["phone"] = "9845012345"
print("Added Phone:", doctor_profile["phone"])

# Delete
del doctor_profile["phone"]
print("After Delete:", doctor_profile)

# ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
# PART 4 - Loop (Python mein Direct!)
# ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

print("\n--- Loop ---")


session = {
    "doctor_name": "Dr. Priya Sharma",
    "patient_name": "Ramesh Kumar",
    "fee": 800,
    "duration": 30
}

# ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
# PART 5 - Spread Operator (Python mein **)
# ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

print("\n--- Merge Dictionaries ---")

basic_info = {
    "name": "Dr. Sharma",
    "id": "DOC001"
}

professional_info = {
    "specialization": "Cardiology",
    "fee": 800
}

# Python mein ** spread operator hai
full_profile = {**basic_info, **professional_info}
print("Merged:", full_profile)


# Python mein ** spread operator hai
full_profile = {**basic_info, **professional_info}
print("Merged:", full_profile)

# Session shuru — doctor busy
active_session = {
    **doctor_profile,
    "available": False,          # Override
    "current_patient": "Ramesh Kumar",
    "session_start": "10:30 AM"
}
print("Available:", active_session["available"])  # False


# ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
# PART 6 - Real Project Use
# ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

doc_name = doctor_profile["name"]
specialization = doctor_profile["specialization"]
doc_fee = doctor_profile["fee"]

print(f"""
╔══════════════════════════════╗
║      VOICEDOC AI SESSION     ║
╠══════════════════════════════╣
║ Doctor : {doc_name}
║ Dept   : {specialization}
║ Fee    : ₹{doc_fee}
║ Patient: {active_session['current_patient']}
║ Time   : {active_session['session_start']}
║ Status : {"Free" if active_session['available'] else "In Session"}
╚══════════════════════════════╝
""")